src/traces/box/calc.js

```diff
@@ -36,112 +36,146 @@ module.exports = function calc(gd, trace) {
         posLetter = 'x';
     }
 
-    val = valAxis.makeCalcdata(trace, valLetter);  // get val
-
-    // size autorange based on all source points
-    // position happens afterward when we know all the pos
-    Axes.expand(valAxis, val, {padded: true});
-
-    // In vertical (horizontal) box plots:
-    // if no x (y) data, use x0 (y0), or name
-    // so if you want one box
-    // per trace, set x0 (y0) to the x (y) value or category for this trace
-    // (or set x (y) to a constant array matching y (x))
-    function getPos(gd, trace, posLetter, posAxis, val) {
-        var pos0;
-        if(posLetter in trace) pos = posAxis.makeCalcdata(trace, posLetter);
-        else {
-            if(posLetter + '0' in trace) pos0 = trace[posLetter + '0'];
-            else if('name' in trace && (
-                        posAxis.type === 'category' ||
-                        (isNumeric(trace.name) &&
-                            ['linear', 'log'].indexOf(posAxis.type) !== -1) ||
-                        (Lib.isDateTime(trace.name) &&
-                         posAxis.type === 'date')
-                    )) {
-                pos0 = trace.name;
-            }
-            else pos0 = gd.numboxes;
-            pos0 = posAxis.d2c(pos0, 0, trace[posLetter + 'calendar']);
-            pos = val.map(function() { return pos0; });
-        }
-        return pos;
-    }
-
-    pos = getPos(gd, trace, posLetter, posAxis, val);
+    var val = valAxis.makeCalcdata(trace, valLetter);
+    var pos = getPos(trace, posLetter, posAxis, val, gd.numboxes);
 
-    // get distinct positions and min difference
     var dv = Lib.distinctVals(pos);
-    posDistinct = dv.vals;
-    dPos = dv.minDiff / 2;
-
-    function binVal(cd, val, pos, posDistinct, dPos) {
-        var posDistinctLength = posDistinct.length,
-            valLength = val.length,
-            valBinned = [],
-            bins = [],
-            i, p, n, v;
-
-        // store distinct pos in cd, find bins, init. valBinned
-        for(i = 0; i < posDistinctLength; ++i) {
-            p = posDistinct[i];
-            cd[i] = {pos: p};
-            bins[i] = p - dPos;
-            valBinned[i] = [];
-        }
-        bins.push(posDistinct[posDistinctLength - 1] + dPos);
-
-        // bin the values
-        for(i = 0; i < valLength; ++i) {
-            v = val[i];
-            if(!isNumeric(v)) continue;
-            n = Lib.findBin(pos[i], bins);
-            if(n >= 0 && n < valLength) valBinned[n].push(v);
+    var posDistinct = dv.vals;
+    var dPos = dv.minDiff / 2;
+    var posBins = makeBins(posDistinct, dPos);
+
+    var vLen = val.length;
+    var pLen = posDistinct.length;
+    var ptsPerBin = initNestedArray(pLen);
+
+    // bin pts info per position bins
+    for(i = 0; i < vLen; i++) {
+        var v = val[i];
+        if(!isNumeric(v)) continue;
+
+        var n = Lib.findBin(pos[i], posBins);
+        if(n >= 0 && n < pLen) {
+            var pt = {v: v, i: i};
```

Comment:
This here was the key to make box point hover and selection somewhat _clean_. 

Box calcdata traces have one item per box to be displayed. Box points data-esque array of objects used to be created during `Box.plot`, now is the structure is setup here. Note that it is important to track the original sample pt index (that `i` above) to convert hovered are selected calcdata items to event data `pointNumber` corresponding to indices in the input `x`/`y` sample arrays

Reply:
... probably best to review this file as a whole:

https://github.com/plotly/plotly.js/blob/6a878b0eef3e52332bd84eca40d024eec1219264/src/traces/box/calc.js

Reply:
Definitely nice and clean - I was curious whether this degraded performance for large data sets, but it looks like very little if at all
```js
function r(n) { var v = 0; for(var i = 0; i < n; i++) v += Math.random(); return v; }
function a(n, m) { var out = new Array(n); for(var i = 0; i < n; i++) out[i] = r(m); return out; }
var y = a(1000000, 10)
// I had to patch timeit to support n = 1
timeit(function() { Plotly.newPlot(gd,[{type: 'box', y: y, jitter: 0.5, hoveron: 'points'}]) }, 1)
```
On my computer 1 million points (of which ~5k are outliers) takes ~3 sec, and difference between this branch and master is within the noise (~10% or less)

Reply:
> (~10% or less)

Thanks for taking a look at this!